pkg/ptrace/ptrace.go

```diff
@@ -0,0 +1,483 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ptrace
+
+/*
+#define _GNU_SOURCE
+#include <sys/wait.h>
+#include <sys/uio.h>
+#include <errno.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdlib.h>
+
+void* Uint64ToPointer(uint64_t addr) {
+	return (void*) addr;
+}
+*/
+import "C"
+
+import (
+	"bytes"
+	"debug/elf"
+	"encoding/binary"
+	"fmt"
+	"io/ioutil"
+	"strconv"
+	"syscall"
+	"unsafe"
+
+	"github.com/pkg/errors"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/pingcap/chaos-mesh/pkg/mapreader"
+)
+
+var log = ctrl.Log.WithName("ptrace")
+
+const waitPidErrorMessage = "waitpid ret value: %d"
+
+// If it's on 64-bit platform, `^uintptr(0)` will get a 64-bit number full of one.
+// After shifting right for 63-bit, only 1 will be left. Than we got 8 here.
+// If it's on 32-bit platform, After shifting nothing will be left. Than we got 4 here.
+const ptrSize = 4 << uintptr(^uintptr(0)>>63)
+
+// TracedProgram is a program traced by ptrace
+type TracedProgram struct {
+	pid     int
+	tids    []int
+	Entries *[]mapreader.Entry
+
+	backupRegs *syscall.PtraceRegs
+	backupCode []byte
+}
+
+// Pid return the pid of traced program
+func (p *TracedProgram) Pid() int {
+	return p.pid
+}
+
+func waitPid(pid int) error {
+	ret := waitpid(pid)
+	if ret == pid {
+		return nil
+	}
+
+	return errors.Errorf(waitPidErrorMessage, ret)
+}
+
+func constructPartialProgram(pid int, tidMap map[int]bool) *TracedProgram {
```

Comment:
I should use better way (such as defer and check ret value) to handle error here.